### ISSUE TITLE
Interpreter broken on master when exiting naturally

### ISSUE THREAD

**Issue by asterite** [crystal-lang/crystal#12769]
This is with Crystal master.

With this file:

```crystal
# foo.cr
p 1
```

Run it with the compiled interpreter:

```
$ bin/crystal i foo.cr
```

Here's part of the output:

```
Crystal interpreter 1.7.0-dev [3973f44d5] (2022-11-16).
EXPERIMENTAL SOFTWARE: if you find a bug, please consider opening an issue in
https://github.com/crystal-lang/crystal/issues/new/
1
Invalid memory access (signal 11) at address 0x25
[0x102900d18] *Exception::CallStack::print_backtrace:Nil +104 in /Users/aryborenszweig/Projects/c
rystal/.build/crystal
[0x1028d7e14] ~procProc(Int32, Pointer(LibC::SiginfoT), Pointer(Void), Nil)@src/signal.cr:127 +320 in /Users/aryborenszweig/Projects/crystal/.build/crystal
[0x1ae18c4e4] _sigtramp +56 in /usr/lib/system/libsystem_platform.dylib
[0x1036e96ac] *Crystal::Repl::Interpreter#interpret<Crystal::ASTNode+, Crystal::Type+>:Crystal::Repl::Value +464004 in /Users/aryborenszweig/Projects/crystal/.build/crystal
...
```

Doing a git bisect, the first bad commit is 3973f44d5, which is this PR: https://github.com/crystal-lang/crystal/pull/12608

I didn't investigate yet why that commit broke things... maybe there's an issue with `arc4random` in interpreted mode.

**Comment by asterite** [comment]
I have no idea what's going on but for now we could avoid using `arc4random` in interpreted mode until we figure it out.

I don't think this is related to arc4random though: if the hasher seed buffer isn't filled at all the same problem happens. No idea why.

**Comment by HertzDevil** [comment]
I tried to do `make clean crystal && git cherry-pick --no-commit 3973f44` and the first bad commit was 4abeeb1, which was #12405. Reverting just that commit on master also unbreaks the interpreter.

**Comment by asterite** [comment]
Interesting. Then I think we should revert it. Such optimizations are good but are worthless if there's no way to do incremental or modular compilation. So it's better that things work correctly.

**Comment by straight-shoota** [comment]
Resolved by #12849
